mcstatus/motd.py

```diff
@@ -0,0 +1,820 @@
+                parsed_motd.append(style_val)
+
+        if (text := item.get("text")) is not None:
+            parsed_motd.extend(cls._parse_as_str(text, bedrock=bedrock))
+        if (translate := item.get("translate")) is not None:
+            parsed_motd.append(TranslationTag(translate))
+        parsed_motd.append(Formatting.RESET)
+
+        if "extra" in item:
+            auto_add = list(filter(lambda e: type(e) is Formatting and e != Formatting.RESET, parsed_motd))
+
+            for element in item["extra"]:
+                parsed_motd.extend(cls._parse_as_dict(element, auto_add=auto_add.copy()))
+
+        return parsed_motd
+
+    @staticmethod
+    def _parse_as_str(raw: str, *, bedrock: bool = False) -> list[ParsedMotdItem]:
+        """Parse a MOTD when it's string.
+
+        .. note:: This method returns a lot of empty strings, use ``Motd.simplified`` to remove them.
+
+        :param raw: Raw MOTD, directly from server.
+        :param bedrock: Is server Bedrock Edition?
+            Ignores ``MinecraftColor.MINECOIN_GOLD`` if it's ``False``.
+        :returns: ``ParsedMotdItem`` list, which need to be passed to ``__init__``.
+        """
+        parsed_motd: list[ParsedMotdItem] = []
+
+        split_raw = MOTD_COLORS_RE.split(raw)
+        for element in split_raw:
+            clean_element = element.lstrip("&§").lower()
+            standardized_element = element.replace("&", "§").lower()
+
+            if standardized_element == "§g" and not bedrock:
+                parsed_motd.append(element)  # minecoin_gold on java server
+                continue
+
+            if standardized_element.startswith("§"):
+                try:
+                    parsed_motd.append(MinecraftColor(clean_element))
+                except ValueError:
+                    try:
+                        parsed_motd.append(Formatting(clean_element))
+                    except ValueError:
+                        # just a text
+                        parsed_motd.append(element)
+            else:
+                parsed_motd.append(element)
+
+        return parsed_motd
+
+    def simplified(self) -> Self:
+        """Simplify MOTD by removing unused formatting, unaffected colors etc.
+
+        :returns: New instance with simplified ``parsed`` field.
+        """
+        parsed = self.parsed.copy()
+        simplifies_class = MotdSimplifies(parsed)
+        to_remove = set()
+
+        for hook in simplifies_class.get_simplifies():
+            to_remove = to_remove.union(hook())
+
+        # To safely remove multiple indices by popping, we need to go from last index
+        # in a sorted order, so that the indices don't change on us.
+        for index in sorted(to_remove, reverse=True):
+            parsed.pop(index)
+
+        return __class__(parsed, self.bedrock)
+
+    @property
+    def _html_transformer(self) -> HtmlTransformer:
+        if self.HTML_TRANSFORMER is None:
+            self.HTML_TRANSFORMER = HtmlTransformer
```

Comment:
> the initialization of these transformer classes should be happening in `__init__`, as initializations can take a lot of time in python, doing so before each call is kind of weird.

`BaseTransformer.__init__` takes Motd instance, so we can't initialize them easily with dataclasses.

> If this is an attempt at saving some memory and time by not having instances of all of these transformer classes stored in the motd class, then these properties should be cached. So that the instance is only created once, when the transformation is first requested, but when it's requested again, it will use the cached already created instance.
>
> Still though, since we treat the transformed values as properties, it implies that they should be very quick to compute. In any other case, we should instead be using functions, to signify that the user should instead store that result, as it's computation may take a while. And in this case, I'm starting to think that functions might be a better idea here with all of the complexity under the hood.

Well, I though firstly that this will be cached properties, but it looks overengineering for me. These classes do only one thing in the `__init__` - `self._motd = motd`, so there is no need to cache them. But also, we should count that some of our users use `mcstatus` to ping thousands of servers in small-time, so I thought to not initialize it on every Motd init. However, I should see how much time it takes, if this will be too low - it should be set in `__init__`. Else - methods.

Also, I do not think that type checkers/linters will like method monkeypatching, but it's completely different with properties.

Reply:
The issue is that this shouldn't be solved by initializing these in `__init__` if they take way too long. Because if we do that, every status request will take longer, and again, that's a problem with people running mcstatus for a lot of servers.

Caching does solve this, sort of, however it's a bit unfair, as most people will probably only call this property once, and yet it will already mean leaving the initialized transformer class stored, taking up memory. Also note that Motd is presumably a mutable class, so we would need to address for changes to the stored motd elements with this caching.

In cases like these, properties are just not the way to go. If it's this expensive to do something, we shouldn't use hacky solutions like caching that has disadvantages elsewhere, without the user ever realising how expensive it might be. Instead, we should absolutely use functions instead. And if the transformer classes should stay, then we should rely purely on them doing all of the work, and not including any of this into the Motd class directly.

If the transformer classes will not stay, we should convert all of these transformations into full functions, rather than properties, if they're really that expensive and require a full class to be initialized (which I still don't really think is justified).

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/5a99136ba6ccfe68f2ad9ac32efeb5742a8ef097..33969fa91ce672da8406b30686151ffb9113ab62).